lib/core/utils/node-serializer.js

```diff
@@ -0,0 +1,40 @@
+import assert from './assert';
+import DqElement from './dq-element';
+
+let serializer = null;
+
+/**
+ * @param {Object} newSerializer
+ * @property {Function} toSpec Converts an Element or VirtualNode to a "spec", a form
+ * suitable for JSON.stringify to consume. Output must include all properties
+ * that DqElement.toJSON() would have. Will always be invoked from the
+ * input element's original page context.
+ * @property {Function} mergeSpecs Merges two specs (produced by toSpec) which
+ * represent element's parent frame and an element, respectively. Will
+ * *not* necessarily be invoked from *either* node's original page context.
+ * This operation must be associative, that is, these two expressions must
+ * produce the same result:
+ * - mergeSpecs(a, mergeSpecs(b, c))
+ * - mergeSpecs(mergeSpecs(a, b), c)
+ */
+export default function nodeSerializer(newSerializer) {
+  assert(typeof serializer === 'object', 'serializer must be an object');
+  serializer = newSerializer;
+}
+
+Object.assign(nodeSerializer, {
```

Comment:
Why wouldn't you be able to type / get intellisense on this? I think this would do it:

```ts
  type NodeToSpec = <T extends SerialDqElement = SerialDqElement>
    (node: VirtualNode | Element) => T;
  type MergeSpecs = <T extends SerialDqElement = SerialDqElement>
    (parentSpec: SerialDqElement, childSpec: SerialDqElement) => T;

  interface Utils {
    nodeSerializer: {
      ({ isSpec, mergeSpecs }: { isSpec?: NodeToSpec, mergeSpecs?: MergeSpecs }): void,
      isSpec: NodeToSpec,
      mergeSpecs: NodeToSpec
    }
```

But sure, this could also just be an object with a "set" function if you want. I like this approach better personally, but it's not a big deal.

Reply:
As discussed, I'm replacing this with a `.update()` method.